File: src/main/java/com/multicampus/gamesungcoding/a11ymarketserver/feature/cart/controller/CartController.java
package com.multicampus.gamesungcoding.a11ymarketserver.feature.cart.controller;


import com.multicampus.gamesungcoding.a11ymarketserver.feature.cart.dto.*;
import com.multicampus.gamesungcoding.a11ymarketserver.feature.cart.service.CartService;
import jakarta.validation.Valid;
import jakarta.validation.constraints.NotNull;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.*;

import java.net.URI;
import java.util.UUID;

@RestController
@RequiredArgsConstructor
@RequestMapping("/api")
@Validated
public class CartController {

    private final CartService cartService;

    // GET /api/v1/cart 목록 조회 기능
    @GetMapping("/v1/cart/me")
    public ResponseEntity<CartItemListResponse> getCart(
            @AuthenticationPrincipal UserDetails userDetails) {

        var userEmail = userDetails.getUsername();
        return ResponseEntity.ok(
                new CartItemListResponse(
                        cartService.getCartItems(userEmail),
                        cartService.getCartTotal(userEmail)
                )
        );
    }

    // POST /api/v1/cart/items 상품 추가 기능
    @PostMapping("/v1/cart/items")
    @ResponseStatus(HttpStatus.CREATED)
    public ResponseEntity<CartItemUpdatedResponse> addItem(
            @Valid @RequestBody CartAddRequest req,
            @AuthenticationPrincipal UserDetails userDetails) {

        return ResponseEntity
                .created(URI.create("/api/v1/cart/me"))
                .body(cartService.addItem(req, userDetails.getUsername()));
    }

    // PATCH /api/v1/cart/items/{cartItemId} 수량 조정 기능
    @PatchMapping("/v1/cart/items/{cartItemId}")
    public ResponseEntity<CartItemUpdatedResponse> updateQuantity(
            @PathVariable @NotNull String cartItemId,
            @Valid @RequestBody CartQtyUpdateRequest body,
            @AuthenticationPrincipal UserDetails userDetails) {

        return ResponseEntity.ok(
                cartService.updateQuantity(
                        UUID.fromString(cartItemId),
                        body.quantity(),
                        userDetails.getUsername())
        );
    }

    // DELETE /api/v1/cart/items 상품 삭제 기능
    @DeleteMapping("/v1/cart/items")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    public ResponseEntity<Void> deleteItems(
            @Valid @RequestBody CartItemDeleteRequest itemIds,
            @AuthenticationPrincipal UserDetails userDetails) {

        cartService.deleteItems(itemIds, userDetails.getUsername());
        return ResponseEntity.noContent().build();
    }
}

File: src/main/kotlin/com/multicampus/gamesungcoding/a11ymarketserver/feature/cart/controller/CartController.kt
package com.multicampus.gamesungcoding.a11ymarketserver.feature.cart.controller

import com.multicampus.gamesungcoding.a11ymarketserver.feature.cart.dto.*
import com.multicampus.gamesungcoding.a11ymarketserver.feature.cart.service.CartService
import jakarta.servlet.http.HttpServletResponse
import jakarta.validation.Valid
import jakarta.validation.constraints.NotNull
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpStatus
import org.springframework.security.core.annotation.AuthenticationPrincipal
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.validation.annotation.Validated
import org.springframework.web.bind.annotation.*
import java.util.*


@RestController
@RequestMapping("/api/v1/cart")
@Validated
class CartController(
    private val cartService: CartService
) {
    // GET /api 목록 조회 기능
    @GetMapping("/me")
    fun getCart(
        @AuthenticationPrincipal userDetails: UserDetails
    ): CartItemListResponse = cartService.getCartItems(userDetails.username)

    @GetMapping("/me/items/count")
    fun getCartItemCount(
        @AuthenticationPrincipal userDetails: UserDetails
    ): CartItemCountResponse = cartService.getCartItemCount(userDetails.username)

    // POST /api/items 상품 추가 기능
    @PostMapping("/me/items")
    @ResponseStatus(HttpStatus.CREATED)
    fun addItem(
        @Valid @RequestBody req: @Valid CartAddRequest,
        @AuthenticationPrincipal userDetails: UserDetails,
        httpResponse: HttpServletResponse
    ): CartItemUpdatedResponse {
        httpResponse.setHeader(HttpHeaders.LOCATION, "/api/me")
        return cartService.addItem(req, userDetails.username)
    }

    // PATCH /api/items/{cartItemId} 수량 조정 기능
    @PatchMapping("/me/items/{cartItemId}")
    fun updateQuantity(
        @PathVariable @NotNull cartItemId: @NotNull String,
        @Valid @RequestBody body: @Valid CartQtyUpdateRequest,
        @AuthenticationPrincipal userDetails: UserDetails
    ): CartItemUpdatedResponse = cartService.updateQuantity(
        UUID.fromString(cartItemId),
        body.quantity,
        userDetails.username
    )

    // DELETE /api/items 상품 삭제 기능
    @DeleteMapping("/me/items")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun deleteItems(
        @Valid @RequestBody itemIds: @Valid CartItemDeleteRequest,
        @AuthenticationPrincipal userDetails: UserDetails
    ) {
        cartService.deleteItems(itemIds, userDetails.username)
    }
}